hack/generator/pkg/astbuilder/func_details.go

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package astbuilder
+
+import (
+	"fmt"
+	"go/ast"
+	"strings"
+)
+
+type FuncDetails struct {
+	ReceiverIdent *ast.Ident
+	ReceiverType  ast.Expr
+	Name          *ast.Ident
+	Comments      []string
+	Params        []*ast.Field
+	Returns       []*ast.Field
+	Body          []ast.Stmt
+}
+
+// NewTestFuncDetails() returns a FuncDetails for a test method
+// Tests require a particular signature, so this makes it simpler to create test functions
+func NewTestFuncDetails(testName string, body ...ast.Stmt) *FuncDetails {
+
+	// Ensure the method name starts with `Test` as required
+	var name string
+	if strings.HasPrefix(testName, "Test") {
+		name = testName
+	} else {
+		name = "Test_" + testName
+	}
+
+	result := &FuncDetails{
+		Name: ast.NewIdent(name),
+		Body: body,
+	}
+
+	result.AddParameter("t",
+		&ast.StarExpr{
+			X: &ast.SelectorExpr{
+				X:   ast.NewIdent("testing"),
+				Sel: ast.NewIdent("T"),
+			}},
+	)
+
+	return result
+}
+
+// DefineFunc defines a function (header, body, etc), like:
+// 	<comment>
+//	func (<receiverIdent> <receiverType>) <name>(<params...>) (<returns...>) {
+// 		<body...>
+//	}
+func DefineFunc(funcDetails *FuncDetails) *ast.FuncDecl {
+
+	// Filter out any nil statements
+	// this helps creation of the funcDetails go simpler
+	var body []ast.Stmt
+	for _, s := range funcDetails.Body {
+		if s != nil {
+			body = append(body, s)
+		}
+	}
+
+	var comment []*ast.Comment
+	if len(funcDetails.Comments) > 0 {
+		funcDetails.Comments[0] = fmt.Sprintf("// %s %s", funcDetails.Name, funcDetails.Comments[0])
+		AddComments(&comment, funcDetails.Comments)
+	}
+
+	result := &ast.FuncDecl{
+		Name: funcDetails.Name,
+		Doc: &ast.CommentGroup{
+			List: comment,
+		},
+		Type: &ast.FuncType{
+			Params: &ast.FieldList{
+				List: funcDetails.Params,
+			},
+			Results: &ast.FieldList{
+				List: funcDetails.Returns,
+			},
+		},
+		Body: &ast.BlockStmt{
+			List: body,
+		},
+	}
+
+	if funcDetails.ReceiverIdent != nil {
```

Comment:
Reviewing just this bit:

I think the only real difference between what you're doing here and my PR #306 is that I have this check as well:
```
        // Safety check that we are making something valid
	if (funcDetails.ReceiverIdent == nil) != (funcDetails.ReceiverType == nil) {
		panic(fmt.Sprintf("ReceiverIdent and ReceiverType must both be specified, or omitted. ReceiverIdent: %q, ReceiverType: %q",
			funcDetails.ReceiverIdent,
			funcDetails.ReceiverType))
	}
```

I think you want this check just to protect against possibly invalid calls? Other than that my PR is basically a subset of yours.

You mentioned wanting to possibly split this PR up a bit to make it smaller - wondering if it would make sense to strip out all of the updates to astbuilder and do them first, and then I could just abandon my PR and we could merge yours?

Reply:
That was a good idea - see PR #313